fw/http_msg.c

```diff
@@ -1313,115 +1329,253 @@ tfw_http_msg_expand_data(TfwMsgIter *it, struct sk_buff **skb_head,
-			const char *stop)
+tfw_http_msg_expand_from_pool(TfwHttpTransIter *mit, TfwPool* pool,
+			      const TfwStr *str)
 {
-	return tfw_h2_msg_rewrite_data_common(mit, str, stop, memcpy_fast);
+	return __tfw_http_msg_expand_from_pool(mit, pool, str, memcpy_fast);
 }
 
 int
-tfw_h2_msg_rewrite_data_lc(TfwHttpTransIter *mit, const TfwStr *str,
-			   const char *stop)
+tfw_http_msg_expand_from_pool_lc(TfwHttpTransIter *mit, TfwPool* pool,
+				 const TfwStr *str)
+{
+	return __tfw_http_msg_expand_from_pool(mit, pool, str, tfw_cstrtolower);
+}
+
+static void
+__tfw_h2_msg_shrink_frags(TfwMsgIter *it, const char *body,
+			  TfwHttpRespCleanup *cleanup)
 {
-	return tfw_h2_msg_rewrite_data_common(mit, str, stop, tfw_cstrtolower);
+	struct skb_shared_info *si = skb_shinfo(it->skb);
+	skb_frag_t *frag = &si->frags[it->frag];
+	char* frag_addr = skb_frag_address(frag);
+	int len;
+	int i;
+
+	/* Delete all fragments from skb. We will use this skb as skb_head. */
+	if (!body) {
+		for (i = 0; i < si->nr_frags; i++) {
+			struct page *pg = skb_frag_page(&si->frags[i]);
+
+			cleanup->pages[i] = compound_head(pg);
+		}
+		len = it->skb->data_len;
+		cleanup->pages_sz = si->nr_frags;
+		si->nr_frags = 0;
+		it->frag = 0;
+		ss_skb_adjust_data_len(it->skb, -len);
+	} else if (frag_addr != body) {
+		/* Start of body it's not start of frag. Shrink frag.*/
+		len = body - frag_addr;
+		/* Add offset and decrease fragment's size */
+		skb_frag_off_add(frag, len);
+		skb_frag_size_sub(frag, len);
+		ss_skb_adjust_data_len(it->skb, -len);
+	}
+}
+
+/*
+ * Delete SKBs and paged fragments related to @resp that contains response
+ * headers. SKBs and fragments will be "unlinked" and placed to @cleanup.
+ * At this point we can't free SKBs, because data that they contain used
+ * as source for message trasformation.
+ */
+void
+tfw_h2_msg_cutoff_headers(TfwHttpResp *resp, TfwHttpRespCleanup* cleanup)
+{
+	char *begin, *end;
+	TfwMsgIter *it = &resp->mit.iter;
+	char* body = TFW_STR_CHUNK(&resp->body, 0)->data;
+	TfwStr *crlf = TFW_STR_LAST(&resp->crlf);
+	char *off = resp->body.len ? body : crlf->data + crlf->len;
+
+	do {
+		unsigned char i;
+		struct sk_buff *skb;
+		struct skb_shared_info *si = skb_shinfo(it->skb);
+
+		if (skb_headlen(it->skb)) {
+			begin = it->skb->data;
+			end = begin + skb_headlen(it->skb);
+
+			if ((begin <= off) && (end >= off)) {
+				it->frag = -1;
+				/* TODO: Handle this case, most likely for low MTU
+				 * it will be implemented as part of #1703 */
+				return;
+			}
+		}
+		for (i = 0; i < si->nr_frags; i++) {
+			skb_frag_t *f = &si->frags[i];
+
+			begin = skb_frag_address(f);
+			end = begin + skb_frag_size(f);
+
+			if ((begin <= off) && (end >= off)) {
+				it->frag = i;
+				__tfw_h2_msg_shrink_frags(it, body, cleanup);
+				goto end;
+			}
+		}
+
+		skb = it->skb;
+		it->skb = it->skb->next;
+		ss_skb_unlink(&it->skb_head, skb);
+		ss_skb_queue_tail(&cleanup->skb_head, skb);
+
+	} while (it->skb != NULL);
```

Comment:
@RomanBelozerov I'm wondering if we have a test, which sends an HTTP response to Tempesta with very small chunks, especially headers, and Tempesta forwards the response to a client via HTTP/2? I believe we have some test for this, but I'm not sure about HTTP/2 client in this scenario...

Reply:
@RomanBelozerov I created https://github.com/tempesta-tech/tempesta-test/issues/431 for this